gtsfm/utils/homography_decomposition.py

```diff
@@ -0,0 +1,387 @@
+from gtsfm.data_association.point3d_initializer import Point3dInitializer, TriangulationParam, TriangulationExitCode
+
+
+def pose_from_homography_matrix(
+    H: np.ndarray,
+    camera_intrinsics_i1: Cal3Bundler,
+    camera_intrinsics_i2: Cal3Bundler,
+    points1: np.ndarray,
+    points2: np.ndarray,
+) -> Tuple[Rot3, Unit3, np.ndarray, np.ndarray]:
+    """Recover the most probable pose from the given homography matrix.
+
+    Args:
+        H: array of shape (3,3)
+        camera_intrinsics_i1: camera 1's intrinsics.
+        camera_intrinsics_i2: camera 2's intrinsics.
+        points1: array of shape (N,2)
+        points2: array of shape (N,2)
+
+    Returns:
+        i2Ri1: relative rotation matrix.
+        i2Ui1: translation direction.
+        n: array of shape (3,) representing plane normal vector.
+        points3D: array of shape (N,3) representing triangulated 3d points.
+    """
+    if points1.shape != points2.shape:
+        raise RuntimeError("Coordinates of 2d correspondences must have the same shape.")
+
+    K1 = camera_intrinsics_i1.K()
+    K2 = camera_intrinsics_i2.K()
+    i2Ri1_cmbs, i2ti1_cmbs, n_cmbs = decompose_homography_matrix(H, K1, K2)
+
+    points3D = []
+    for i in range(len(i2Ri1_cmbs)):
+        points3D_cmb = check_cheirality(
+            i2Ri1_cmbs[i], i2ti1_cmbs[i], camera_intrinsics_i1, camera_intrinsics_i2, points1, points2
+        )
+        print(f"hypothesis {i}  -> {len(points3D_cmb)}")
+        print(i2ti1_cmbs[i])
+        if len(points3D_cmb) >= len(points3D):
+            i2Ri1 = i2Ri1_cmbs[i]
+            i2ti1 = i2ti1_cmbs[i]
+            n = n_cmbs[i]
+            points3D = points3D_cmb
+
+    points3D = np.array(points3D)
+
+    # Note: if norm(i2ti1) == 0, this is a pure rotation. Otherwise, planar scene.
+
+    # Note: we cannot blindly cast the translation direction to Unit3, as zero translation is valid
+    # from a rotation, and should be noted in 1dsfm.
+    return i2Ri1, Unit3(i2ti1), n, points3D
+
+
+def check_cheirality(
+    i2Ri1: Rot3,
+    i2ti1: np.ndarray,
+    camera_intrinsics_i1: Cal3Bundler,
+    camera_intrinsics_i2: Cal3Bundler,
+    points1: np.ndarray,
+    points2: np.ndarray,
+) -> np.ndarray:
+    """
+    Args:
+        R: array of shape (3,3)
+        t: array of shape (3,)
+        points1:
+        points2:
+
+    Returns:
+        points3D: array of shape (N,3)
+    """
+    if points1.shape != points2.shape:
+        raise RuntimeError("Coordinates of 2d correspondences must have the same shape.")
+
+    i2Ti1 = Pose3(i2Ri1, i2ti1)
+    camera_dict = {
+        1: PinholeCameraCal3Bundler(i2Ti1, camera_intrinsics_i1),
+        2: PinholeCameraCal3Bundler(Pose3(), camera_intrinsics_i2),
+    }
+
+    # set the camera 2 pose as the global coordinate system
+    # COLMAP states that wTc for camera 2 is equal to (R,t) from homography.
+    # given wTi1 and wTi2, then if
+    # i2Ti1 = wTi2.inverse() * wTi1
+    # thus, wTi2 is identity, and wTi1 = i2Ti1
```

Comment:
that is just simple math. no need to cite COLMAP there. there is probably also no need for this comment. But if you'd like, a more useful comment would be: camera poses are in a "local world" frame at i2.

Reply:
also, if there is need for a comment, it should be moved before the dict.